demos/image_processing_demo/cpp/visualizer.cpp

```diff
@@ -0,0 +1,152 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#include <iostream>
+#include <set>
+#include <string>
+#include "visualizer.hpp"
+
+Visualizer::Visualizer() {
+    cv::namedWindow(winName, cv::WINDOW_AUTOSIZE);
+    mode = "result";
+    inputImg = cv::Mat(resolution, CV_32FC3, 0.);
+    resultImg = cv::Mat(resolution, CV_32FC3, 0.);
+    displayImg = cv::Mat(resolution, CV_32FC3, 0.);
+    trShown = false;
+    isSetResolution = false;
+}
+
+cv::Size Visualizer::getSize() {
+    return resolution;
+}
+
+void Visualizer::handleKey(int key) {
+    key = std::tolower(key);
+    if (key == 'a') {
+        helpShown = 1 - helpShown;
+    }
+    if (key == 'o') {
+        mode = "orig";
+        addTrackbar();
+    }
+    if (key == 'v') {
+        mode = "diff";
+        addTrackbar();
+    }
+    if (key == 'r') {
+        mode = "result";
+        disableTrackbar();
+    }
+}
+
+cv::Mat Visualizer::renderResultData(ImageResult result, OutputTransform& transform) {
+    if (!result.metaData) {
+        throw std::invalid_argument("Renderer: metadata is null");
+    }
+    // Input image is stored inside metadata, as we put it there during submission stage
+    inputImg = result.metaData->asRef<ImageMetaData>().img;
+    if (inputImg.empty()) {
+        throw std::invalid_argument("Renderer: image provided in metadata is empty");
+    }
+
+    if (!isSetResolution) {
+        transform.resize(result.resultImage);
+        setResolution(transform);
+    }
+
+    cv::resize(result.resultImage, result.resultImage, resolution);
+    cv::resize(inputImg, inputImg, resolution);
+
+    if (inputImg.channels() != result.resultImage.channels()){
+        cv::cvtColor(result.resultImage, resultImg, cv::COLOR_GRAY2BGR);
+    }
+    else
+        resultImg = result.resultImage;
+    changeDisplayImg();
+    return displayImg;
+}
+
+void Visualizer::show(cv::Mat img) {
+    if (img.empty()) {
+        changeDisplayImg();
+        img = displayImg;
+    }
+
+    if (helpShown) {
+        float pad = 10;
+        float margin = 40;
+        std::vector<std::string> lines;
+        std::string line;
+        std::istringstream textStream(helpMessage);
+        while (std::getline(textStream, line, '\n'))
+           lines.push_back(line);
+        int baseline = 0;
+        int lineH = cv::getTextSize(lines[0], cv::FONT_HERSHEY_COMPLEX, 0.75, 1, &baseline).height + pad;
+        for (size_t i = 0; i < lines.size(); ++i) {
+            cv::putText(img, lines[i], cv::Point(pad, margin + baseline + (i + 1)*lineH),  cv::FONT_HERSHEY_COMPLEX,
+                        0.75, cv::Scalar(255, 0, 255));
+        }
+    }
+    cv::imshow(winName, img);
+}
+
+void Visualizer::changeDisplayImg() {
+    if (mode == "orig") {
+        displayImg = inputImg.clone();
+        resultImg(cv::Rect(0, 0, slider, inputImg.rows)).copyTo(displayImg(cv::Rect(0, 0, slider, resultImg.rows)));
```

Comment:
This line fails on Windows when O is pressed: `slider` still has 0. Probably because `cv::createTrackbar()` does its stuff only when `cv::waitKey()` is called.

Reply:
fixed